File: windows/TZReader.Core/BookReport.cs
using System.Text;
using System.Text.RegularExpressions;
using System.Xml.Linq;

namespace TZReader.Core;

/// <summary>
/// 書籍を開いたときに何が起きたかを、決定的な値だけで要約する。
/// アプリの診断画面と、実装間の突き合わせの両方で使う。
/// 環境やタイミングに依存する値（時刻、所要時間、絶対パス）は入れない。
/// </summary>
public sealed record BookReport
{
    public required string Format { get; init; }
    public required string Layout { get; init; }
    public required string Direction { get; init; }
    public required int SpineCount { get; init; }
    public required int TocEntryCount { get; init; }
    public required int TocMaxDepth { get; init; }
    public required bool HasCover { get; init; }

    public required IReadOnlyList<string> MissingResources { get; init; }
    public required IReadOnlyList<string> MissingTocTargets { get; init; }
    public required IReadOnlyList<string> MissingSpineItems { get; init; }

    public required int CssFileCount { get; init; }
    public required int NonUtf8CssCount { get; init; }
    public required int LegacyCssFileCount { get; init; }
    public required IReadOnlyList<CssCompat.Change> CssChanges { get; init; }

    public required int XhtmlCount { get; init; }
    public required int MalformedXhtmlCount { get; init; }

    public required int ImageCount { get; init; }
    public required int FontCount { get; init; }
    public required int StoredEntryCount { get; init; }
    public required int DeflatedEntryCount { get; init; }
    public required bool HasEncryptionMetadata { get; init; }

    public bool IsHealthy =>
        MissingResources.Count == 0 && MissingTocTargets.Count == 0
        && MissingSpineItems.Count == 0 && MalformedXhtmlCount == 0;

    public static BookReport Make(EpubArchive archive, EpubPublication publication)
    {
        var missingResources = new List<string>();
        var missingSpine = new List<string>();
        var cssFiles = 0;
        var nonUtf8 = 0;
        var legacyCss = 0;
        var changeTotals = new Dictionary<string, CssCompat.Change>(StringComparer.Ordinal);
        var xhtml = 0;
        var malformed = 0;
        var images = 0;
        var fonts = 0;
        var stored = 0;
        var deflated = 0;

        foreach (var name in archive.Names.OrderBy(n => n, StringComparer.Ordinal))
        {
            if (archive.IsStored(name))
            {
                stored++;
            }
            else
            {
                deflated++;
            }

            switch (Extension(name))
            {
                case "css":
                    cssFiles++;
                    byte[] cssData;
                    try
                    {
                        cssData = archive.Read(name);
                    }
                    catch (Exception)
                    {
                        continue;
                    }
                    if (!IsValidUtf8(cssData))
                    {
                        nonUtf8++;
                    }
                    var cssText = CssCompat.DecodeText(cssData);
                    if (!cssText.Contains("-epub-", StringComparison.Ordinal))
                    {
                        continue;
                    }
                    legacyCss++;
                    foreach (var change in CssCompat.Rewrite(cssText).Changes)
                    {
                        var key = change.From + "→" + change.To;
                        changeTotals[key] = changeTotals.TryGetValue(key, out var existing)
                            ? existing with { Count = existing.Count + change.Count }
                            : change;
                    }
                    break;

                case "xhtml":
                case "html":
                case "htm":
                    xhtml++;
                    try
                    {
                        var text = CssCompat.DecodeText(archive.Read(name));
                        var normalized = CssCompat.RewriteXhtml(text).Css;
                        if (!ParsesAsXml(normalized))
                        {
                            malformed++;
                        }
                    }
                    catch (Exception)
                    {
                        malformed++;
                    }
                    break;

                case "png" or "jpg" or "jpeg" or "gif" or "svg" or "webp":
                    images++;
                    break;

                case "ttf" or "otf" or "woff" or "woff2":
                    fonts++;
                    break;
            }
        }

        foreach (var link in publication.ReadingOrder.Where(l => !archive.Contains(l.Href)))
        {
            missingSpine.Add(link.Href);
        }

        var tocTargets = new List<string>();
        var maxDepth = 0;
        var entryCount = 0;

        void Walk(IReadOnlyList<TocEntry> entries, int depth)
        {
            if (entries.Count == 0)
            {
                return;
            }
            maxDepth = Math.Max(maxDepth, depth);
            foreach (var entry in entries)
            {
                entryCount++;
                if (entry.Href is not null && !archive.Contains(entry.Href))
                {
                    tocTargets.Add(entry.Href);
                }
                Walk(entry.Children, depth + 1);
            }
        }
        Walk(publication.TableOfContents, 1);

        // 章から参照されている画像や CSS のうち、実体が無いものを拾う。
        foreach (var link in publication.ReadingOrder)
        {
            byte[] data;
            try
            {
                data = archive.Read(link.Href);
            }
            catch (Exception)
            {
                continue;
            }
            var basePath = EpubParser.DirectoryOf(link.Href);
            foreach (var reference in References(CssCompat.DecodeText(data)))
            {
                var resolved = Paths.Resolve(basePath, reference);
                if (resolved.Length > 0 && !archive.Contains(resolved) && !missingResources.Contains(resolved))
                {
                    missingResources.Add(resolved);
                }
            }
        }

        return new BookReport
        {
            Format = publication.Layout == PublicationLayout.Fixed ? "fixedEPUB" : "reflowableEPUB",
            Layout = publication.Layout == PublicationLayout.Fixed ? "fixed" : "reflowable",
            Direction = publication.Direction == ReadingDirection.Rtl ? "rtl" : "ltr",
            SpineCount = publication.ReadingOrder.Count,
            TocEntryCount = entryCount,
            TocMaxDepth = maxDepth,
            HasCover = publication.CoverHref is not null,
            MissingResources = missingResources.Order(StringComparer.Ordinal).ToList(),
            MissingTocTargets = tocTargets.Distinct(StringComparer.Ordinal).Order(StringComparer.Ordinal).ToList(),
            MissingSpineItems = missingSpine.Order(StringComparer.Ordinal).ToList(),
            CssFileCount = cssFiles,
            NonUtf8CssCount = nonUtf8,
            LegacyCssFileCount = legacyCss,
            CssChanges = changeTotals.Values
                .OrderBy(c => c.From, StringComparer.Ordinal)
                .ThenBy(c => c.To, StringComparer.Ordinal)
                .ToList(),
            XhtmlCount = xhtml,
            MalformedXhtmlCount = malformed,
            ImageCount = images,
            FontCount = fonts,
            StoredEntryCount = stored,
            DeflatedEntryCount = deflated,
            HasEncryptionMetadata = archive.Contains("META-INF/encryption.xml"),
        };
    }

    /// <summary>章が参照している src / href のうち、外部 URL と fragment 以外を返す。</summary>
    private static IEnumerable<string> References(string html)
    {
        foreach (Match match in Regex.Matches(
            html, @"(?:src|href|xlink:href)\s*=\s*[""']([^""']+)[""']", RegexOptions.IgnoreCase))
        {
            var raw = match.Groups[1].Value.Trim();
            if (raw.Length == 0 || raw.StartsWith('#') || raw.StartsWith("data:", StringComparison.Ordinal))
            {
                continue;
            }
            if (raw.Contains("://", StringComparison.Ordinal) || raw.StartsWith("mailto:", StringComparison.Ordinal))
            {
                continue;
            }
            var path = Paths.StripFragment(raw).Path;
            if (path.Length > 0)
            {
                yield return path;
            }
        }
    }

    private static string Extension(string name)
    {
        var index = name.LastIndexOf('.');
        return index < 0 ? string.Empty : name[(index + 1)..].ToLowerInvariant();
    }

    private static bool IsValidUtf8(byte[] data)
    {
        try
        {
            new UTF8Encoding(false, throwOnInvalidBytes: true).GetString(data);
            return true;
        }
        catch (DecoderFallbackException)
        {
            return false;
        }
    }

    internal static bool ParsesAsXml(string text)
    {
        try
        {
            using var reader = System.Xml.XmlReader.Create(
                new StringReader(text),
                new System.Xml.XmlReaderSettings
                {
                    DtdProcessing = System.Xml.DtdProcessing.Ignore,
                    XmlResolver = null,
                });
            XDocument.Load(reader);
            return true;
        }
        catch (Exception)
        {
            return false;
        }
    }
}

File: macos/Sources/ChoroReader/Document/BookReport.swift
import Foundation

/// 書籍を開いたときに何が起きたかを、決定的な値だけで要約する。
///
/// 用途は 2 つある。
/// 1. アプリの診断画面：表示がおかしいときに、原因が書籍側か実装側かを切り分ける。
/// 2. 実装間の突き合わせ：同じ書籍に対して Swift 版と Windows 版が同じレポートを返すことを確かめる。
/// どちらにも使えるよう、環境やタイミングに依存する値（時刻、所要時間、絶対パス）は入れない。
struct BookReport: Codable, Hashable {
    var format: String
    var layout: String
    var direction: String
    var spineCount: Int
    var tocEntryCount: Int
    var tocMaxDepth: Int
    var hasCover: Bool

    /// manifest が指しているのにアーカイブに無いリソース。壊れた EPUB の主要な兆候。
    var missingResources: [String]
    /// 目次の参照先のうち、アーカイブに無いもの。
    var missingTOCTargets: [String]
    /// spine にあるのにアーカイブに無い章。
    var missingSpineItems: [String]

    var cssFileCount: Int
    /// UTF-8 として読めなかった CSS。文字コード正規化の対象。
    var nonUTF8CSSCount: Int
    /// `-epub-` プレフィックスを含む CSS の数と、変換の内訳。
    var legacyCSSFileCount: Int
    var cssChanges: [CSSCompat.Change]

    var xhtmlCount: Int
    /// XML として解釈できない章の数。application/xhtml+xml では表示できず、text/html で配信する対象。
    var malformedXHTMLCount: Int

    var imageCount: Int
    var fontCount: Int
    var storedEntryCount: Int
    var deflatedEntryCount: Int
    /// META-INF/encryption.xml の有無。難読化フォントや DRM の判別に使う。
    var hasEncryptionMetadata: Bool

    var isHealthy: Bool {
        missingResources.isEmpty && missingTOCTargets.isEmpty && missingSpineItems.isEmpty
            && malformedXHTMLCount == 0
    }

    static func make(archive: ZipArchive, publication: EPUBPublication) -> BookReport {
        var missingResources: [String] = []
        var missingSpine: [String] = []
        var cssFiles = 0
        var nonUTF8 = 0
        var legacyCSS = 0
        var changeTotals: [String: CSSCompat.Change] = [:]
        var xhtml = 0
        var malformed = 0
        var images = 0
        var fonts = 0
        var stored = 0
        var deflated = 0

        for name in archive.names.sorted() {
            guard let entry = archive.entries[name] else { continue }
            if entry.method == 0 { stored += 1 } else { deflated += 1 }

            switch (name as NSString).pathExtension.lowercased() {
            case "css":
                cssFiles += 1
                guard let data = try? archive.read(name) else { continue }
                if String(data: data, encoding: .utf8) == nil { nonUTF8 += 1 }
                let text = CSSCompat.decodeText(data)
                guard text.contains("-epub-") else { continue }
                legacyCSS += 1
                for change in CSSCompat.rewrite(css: text).changes {
                    let key = change.from + "→" + change.to
                    if var existing = changeTotals[key] {
                        existing.count += change.count
                        changeTotals[key] = existing
                    } else {
                        changeTotals[key] = change
                    }
                }
            case "xhtml", "html", "htm":
                xhtml += 1
                guard let data = try? archive.read(name) else { continue }
                let normalized = Data(CSSCompat.rewriteXHTML(CSSCompat.decodeText(data)).css.utf8)
                if (try? XMLDocument(data: normalized, options: [.nodeLoadExternalEntitiesNever])) == nil {
                    malformed += 1
                }
            case "png", "jpg", "jpeg", "gif", "svg", "webp":
                images += 1
            case "ttf", "otf", "woff", "woff2":
                fonts += 1
            default:
                break
            }
        }

        for link in publication.readingOrder where !archive.contains(link.href) {
            missingSpine.append(link.href)
        }

        var tocTargets: [String] = []
        var maxDepth = 0
        var entryCount = 0
        func walk(_ entries: [TOCEntry], depth: Int) {
            guard !entries.isEmpty else { return }
            maxDepth = max(maxDepth, depth)
            for entry in entries {
                entryCount += 1
                if let href = entry.href, !archive.contains(href) { tocTargets.append(href) }
                walk(entry.children, depth: depth + 1)
            }
        }
        walk(publication.tableOfContents, depth: 1)

        // 章から参照されている画像や CSS のうち、実体が無いものを拾う。
        for link in publication.readingOrder {
            guard let data = try? archive.read(link.href) else { continue }
            let base = (link.href as NSString).deletingLastPathComponent
            for reference in references(in: CSSCompat.decodeText(data)) {
                let resolved = EPUBParser.resolve(base: base, href: reference)
                if !resolved.isEmpty, !archive.contains(resolved), !missingResources.contains(resolved) {
                    missingResources.append(resolved)
                }
            }
        }

        return BookReport(
            format: publication.layout == .fixed ? "fixedEPUB" : "reflowableEPUB",
            layout: publication.layout.rawValue,
            direction: publication.direction.rawValue,
            spineCount: publication.readingOrder.count,
            tocEntryCount: entryCount,
            tocMaxDepth: maxDepth,
            hasCover: publication.coverHref != nil,
            missingResources: missingResources.sorted(),
            missingTOCTargets: Array(Set(tocTargets)).sorted(),
            missingSpineItems: missingSpine.sorted(),
            cssFileCount: cssFiles,
            nonUTF8CSSCount: nonUTF8,
            legacyCSSFileCount: legacyCSS,
            cssChanges: changeTotals.values.sorted { ($0.from, $0.to) < ($1.from, $1.to) },
            xhtmlCount: xhtml,
            malformedXHTMLCount: malformed,
            imageCount: images,
            fontCount: fonts,
            storedEntryCount: stored,
            deflatedEntryCount: deflated,
            hasEncryptionMetadata: archive.contains("META-INF/encryption.xml")
        )
    }

    /// 章が参照している src / href のうち、外部 URL と fragment 以外を返す。
    private static func references(in html: String) -> [String] {
        guard let re = try? NSRegularExpression(
            pattern: #"(?i)(?:src|href|xlink:href)\s*=\s*["']([^"']+)["']"#) else { return [] }
        var out: [String] = []
        for match in re.matches(in: html, range: NSRange(html.startIndex..., in: html)) {
            guard let range = Range(match.range(at: 1), in: html) else { continue }
            let raw = String(html[range]).trimmingCharacters(in: .whitespaces)
            if raw.isEmpty || raw.hasPrefix("#") || raw.hasPrefix("data:") { continue }
            if raw.contains("://") || raw.hasPrefix("mailto:") { continue }
            out.append(EPUBParser.stripFragment(raw).path)
        }
        return out.filter { !$0.isEmpty }
    }

    /// 画面表示用の要約。
    var summaryLines: [(label: String, value: String, warning: Bool)] {
        [
            ("形式", format, false),
            ("綴じ方向", direction == "rtl" ? "右開き" : "左開き", false),
            ("章数", "\(spineCount)", false),
            ("目次項目", "\(tocEntryCount)（最大 \(tocMaxDepth) 階層）", false),
            ("表紙", hasCover ? "あり" : "なし", false),
            ("CSS", "\(cssFileCount) 件（うち旧記法 \(legacyCSSFileCount)、非 UTF-8 \(nonUTF8CSSCount)）",
             legacyCSSFileCount > 0 || nonUTF8CSSCount > 0),
            ("章の XHTML", "\(xhtmlCount) 件（XML として不正 \(malformedXHTMLCount)）", malformedXHTMLCount > 0),
            ("画像 / フォント", "\(imageCount) / \(fontCount)", false),
            ("ZIP エントリ", "無圧縮 \(storedEntryCount) / 圧縮 \(deflatedEntryCount)", false),
            ("欠落リソース", missingResources.isEmpty ? "なし" : "\(missingResources.count) 件",
             !missingResources.isEmpty),
            ("目次の参照先欠落", missingTOCTargets.isEmpty ? "なし" : "\(missingTOCTargets.count) 件",
             !missingTOCTargets.isEmpty),
            ("暗号化メタデータ", hasEncryptionMetadata ? "あり" : "なし", hasEncryptionMetadata),
        ]
    }
}
